openldap/2.6/debian-11/rootfs/opt/bitnami/scripts/libopenldap.sh

```diff
@@ -0,0 +1,1129 @@
+        fi
+        if [[ -z "$LDAP_TLS_CA_FILE" ]]; then
+            print_validation_error "You must provide a CA X.509 certificate in order to use TLS"
+        elif [[ ! -f "$LDAP_TLS_CA_FILE" ]]; then
+            print_validation_error "The CA X.509 certificate file in the specified path ${LDAP_TLS_CA_FILE} does not exist"
+        fi
+    fi
+
+    read -r -a users <<< "$(tr ',;' ' ' <<< "${LDAP_USERS}")"
+    read -r -a passwords <<< "$(tr ',;' ' ' <<< "${LDAP_PASSWORDS}")"
+    if [[ "${#users[@]}" -ne "${#passwords[@]}" ]]; then
+        print_validation_error "Specify the same number of passwords on LDAP_PASSWORDS as the number of users on LDAP_USERS!"
+    fi
+
+    if [[ -n "$LDAP_PORT_NUMBER" ]] && [[ -n "$LDAP_LDAPS_PORT_NUMBER" ]]; then
+        if [[ "$LDAP_PORT_NUMBER" -eq "$LDAP_LDAPS_PORT_NUMBER" ]]; then
+            print_validation_error "LDAP_PORT_NUMBER and LDAP_LDAPS_PORT_NUMBER are bound to the same port!"
+        fi
+    fi
+
+    [[ "$error_code" -eq 0 ]] || exit "$error_code"
+}
+
+########################
+# Check if OpenLDAP is running
+# Globals:
+#   LDAP_PID_FILE
+# Arguments:
+#   None
+# Returns:
+#   Whether slapd is running
+#########################
+is_ldap_running() {
+    local pid
+    pid="$(get_pid_from_file "${LDAP_PID_FILE}")"
+    if [[ -n "${pid}" ]]; then
+        is_service_running "${pid}"
+    else
+        return 1
+    fi
+}
+
+########################
+# Check if OpenLDAP is not running
+# Arguments:
+#   None
+# Returns:
+#   Whether slapd is not running
+#########################
+is_ldap_not_running() {
+    ! is_ldap_running
+}
+
+########################
+# Start OpenLDAP server in background
+# Arguments:
+#   None
+# Returns:
+#   None
+#########################
+ldap_start_bg() {
+    local -r retries="${1:-12}"
+    local -r sleep_time="${2:-1}"
+    local -a flags=("-h" "ldap://:${LDAP_PORT_NUMBER}/ ${LDAP_LDAPI_URI} " "-F" "${LDAP_CONF_DIR}/slapd.d" "-d" "$LDAP_LOGLEVEL")
+
+    if is_ldap_not_running; then
+        info "Starting OpenLDAP server in background"
+        ensure_dir_exists $(dirname "$LDAP_PID_FILE") ${LDAP_DAEMON_USER} ${LDAP_DAEMON_GROUP}
+        ulimit -n "$LDAP_ULIMIT_NOFILES"
+        am_i_root && flags=("-u" "$LDAP_DAEMON_USER" "${flags[@]}")
+        debug_execute slapd "${flags[@]}" &
+        if ! retry_while is_ldap_running "$retries" "$sleep_time"; then
+            error "OpenLDAP failed to start"
+            return 1
+        fi
+    fi
+}
+
+########################
+# Stop OpenLDAP server
+# Arguments:
+#   $1 - max retries. Default: 12
+#   $2 - sleep between retries (in seconds). Default: 1
+# Returns:
+#   None
+#########################
+ldap_stop() {
+    local -r retries="${1:-12}"
+    local -r sleep_time="${2:-1}"
+
+    are_db_files_locked() {
+        local return_value=0
+        read -r -a db_files <<< "$(find "$LDAP_DATA_DIR" -type f -print0 | xargs -0)"
```

Comment:
why not just use `read -d ''` instead of this strange usage of xargs which exposes you to any of the escaping issues you might have been trying to manage here?